aips/aip-xx-enums.md

```diff
@@ -0,0 +1,253 @@
+            *r = VersionedData::V2{name, age: 0}
+        }
+        _ => {
+            // Already V2, nothing to do
+        }
+    }
+}
+```
+
+It is also possible -- in difference to ADT implementations in other languages -- to perform field selection on shared fields, as in
+
+```move
+fun get_name(data: &VersionedData): &String {
+    data.name
+}
+```
+
+This allows old code to continue to work without changes on features shared between enum variants.
+
+This AIP focuses on the extensions to the Move VM needed to implement enum types. The extensions are discussed in terms of the [Move Binary Format](https://github.com/aptos-labs/aptos-core/blob/main/third_party/move/move-binary-format/src/file_format.rs), including the new opcodes and their semantics. The implementation is in [PR #13812](https://github.com/aptos-labs/aptos-core/pull/13812).
+
+
+
+### Out of Scope
+
+- The actual source language representation of enum types. The description is found in the Move Book. TODO: link
+- Possible future extensions, specifically for optimization purposes. For example, currently variant matching must be compiled to predicates, but a 'switch' instruction with a jump table is more efficient for large match expressions. This instruction may be added later.
+- Dealing with public enum types (or more general, struct types). Extensions to the binary format for this purpose are out of scope for this AIP, however, taken in consideration. 
+
+## High-level Overview
+
+Enum types are interpreted as a special form of struct types. Instead of just fields, the struct layout can be either `Singleton` for a traditional struct or `Variants` for a list of variants. For this reason, the technical name for enum types in the VM is *variant struct*.
+
+There are many advantages of folding enums into the existing struct representation. Specifically since backends consume serialized representations of structs, and don’t care whether a blob is an enum or not, the enum extension does not cascade into those components. Also, for RAM representation of enums this viewpoint is suitable. Like in a C union, the variant tag determines the interpretation of the memory blob, but otherwise it behaves like a struct.
+
+A number of new instructions need to be introduced to deal with enums, notably a new form of borrow operation on fields of variants that aborts for unexpected variants. The later one is needed for allowing frontends to generate matching code via references to the data.
+
+## Impact
+
+This is a significant extension to the Move VM which will highly improve the fidelity of the code which can be written in Move. This first version contains the minimal necessary set of new features, more are needed in the future for improving performance.
+
+## Alternative solutions
+
+There are no viable alternatives known. The closest would be to represent a variant type as a struct with optional fields, but this is inefficient and would not give upgradability. 
+
```

Comment:
should we add something like:

If we think about upgradeability problem itself, serialization protocol could be changed/extended, to support upgradable structs. But that a significant change to the serialization protocol, and enum's solve variety of other needs as well.

Reply:
Added the paragraph